apis/r/R/SOMAExperimentAxisQuery.R

```diff
@@ -163,6 +168,527 @@ SOMAExperimentAxisQuery <- R6::R6Class(
+    #'
+    #' @param X_layers A named character of X layers to add to the Seurat assay;
+    #' names should be one of:
+    #' \itemize{
+    #'  \item \dQuote{\code{counts}} to add the layer as \code{counts}
+    #'  \item \dQuote{\code{data}} to add the layer as \code{data}
+    #'  \item \dQuote{\code{scale.data}} to add the layer as \code{scale.data}
+    #' }
+    #' At least one of \dQuote{\code{counts}} or \dQuote{\code{data}} is required
+    #' @param obs_index Name of column in \code{obs} to add as cell names
+    #' @param var_index Name of column in \code{var} to add as feature names
+    #' @param var_column_names Names of columns in \code{var} to add as
+    #' feature-level meta data
+    #'
+    #' @return An \code{\link[SeuratObject]{Assay}} object
+    #'
+    to_seurat_assay = function(
+      X_layers = c(counts = 'counts', data = 'logcounts'),
+      obs_index = NULL,
+      var_index = NULL,
+      var_column_names = NULL
+    ) {
+      version <- 'v3'
+      .check_seurat_installed()
+      stopifnot(
+        "'X_layers' must be a named character vector" = is.character(X_layers) &&
+          is_named(X_layers, allow_empty = FALSE),
+        "'version' must be a single character value" = is_scalar_character(version),
+        "'obs_index' must be a single character value" = is.null(obs_index) ||
+          (is_scalar_character(obs_index) && !is.na(obs_index)),
+        "'var_index' must be a single character value" = is.null(var_index) ||
+          (is_scalar_character(var_index) && !is.na(var_index)),
+        "'var_column_names' must be a character vector" = is.null(var_column_names) ||
+          is.character(var_column_names) ||
+          is_scalar_logical(var_column_names)
+      )
+      match.arg(version, choices = 'v3')
+      features <- if (is.null(var_index)) {
+        paste0('feature', self$var_joinids())
+      } else {
+        var_index <- match.arg(
+          arg = var_index,
+          choices = self$var_df$attrnames()
+        )
+        self$var(var_index)$GetColumnByName(var_index)$as_vector()
+      }
+      cells <- if (is.null(obs_index)) {
+        paste0('cell', self$obs_joinids())
+      } else {
+        obs_index <- match.arg(
+          arg = obs_index,
+          choices = self$obs_df$attrnames()
+        )
+        self$obs(obs_index)$GetColumnByName(obs_index)$as_vector()
+      }
+      # Check the layers
+      layers <- X_layers[X_layers %in% self$ms$X$names()]
+      if (!length(layers)) {
+        stop("None of the requested X_layers can be found", call. = FALSE)
+      } else if (length(layers) != length(X_layers)) {
+        warning(
+          paste(
+            strwrap(paste(
+              "The following layers cannot be found in this ExperimentQuery:",
+              paste(sQuote(setdiff(x = X_layers, y = layers)), collapse = ', ')
+            )),
+            collapse = '\n'
+          ),
+          call. = FALSE,
+          immediate. = TRUE
+        )
+      }
```

Comment:
Typically in Seurat, we try to persevere for small issues and only error out if something actually cannot be done. In this case, if two layers were passed but one is non existent, Seurat would typically throw a warning and continue on. Seurat would only error if absolutely nothing could be done (eg. no layers existing)

In the (hopefully near) future, I want to add support for appending to existing `Seurat` objects, which would allow the user to make mistakes and correct them later on